File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
  const levels = [];
  let queue = [];
  
  root && queue.push([root, 0]);
  
  while (queue.length) {
    // handling an entire level at once alleviates the need for queue.shift()
    // which is an O(n) operation in JS/TS
    const children = [];
    
    levels.push([])
    
    queue.forEach(([node, depth]) => {
      levels[depth].push(node.val);
      node.left && children.push([node.left, depth + 1]);
      node.right && children.push([node.right, depth + 1]);
    });
    
    queue = children;
  }
  
  return levels;
};

File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type depthNode struct {
    node  *TreeNode
    depth int
}

func levelOrder(root *TreeNode) (levels [][]int) {
    if root == nil {
        return
    }
    
    queue := []depthNode{ depthNode{ node: root, depth: 0 } }
    
    for len(queue) != 0 {
        node, depth := queue[0].node, queue[0].depth
        queue = queue[1:] // dequeue
        
        if len(levels) == depth {
            levels = append(levels, []int{})
        }
        levels[depth] = append(levels[depth], node.Val)
        
        if node.Left != nil {
            queue = append(queue, depthNode{ node.Left, depth + 1 })
        }
        if node.Right != nil {
            queue = append(queue, depthNode{ node.Right, depth + 1 })
        }
    }
    
    return
}